File: src/main/java/com/github/dkoval/leetcode/challenge/NonOverlappingIntervals.java
package com.github.dkoval.leetcode.challenge;

import java.util.Arrays;
import java.util.Comparator;

/**
 * <a href="https://leetcode.com/problems/non-overlapping-intervals/">Non-overlapping Intervals</a>
 * <p>
 * Given an array of intervals intervals where intervals[i] = [starti, endi], return the minimum number of intervals
 * you need to remove to make the rest of the intervals non-overlapping.
 * <p>
 * Constraints:
 * <ul>
 *  <li>1 <= intervals.length <= 10^5</li>
 *  <li>intervals[i].length == 2</li>
 *  <li>-5 * 10^4 <= starti < endi <= 5 * 10^4</li>
 * </ul>
 */
public interface NonOverlappingIntervals {

    int eraseOverlapIntervals(int[][] intervals);

    // O(N*logN) time | O(1) space
    class NonOverlappingIntervalsRev1 implements NonOverlappingIntervals {

        @Override
        public int eraseOverlapIntervals(int[][] intervals) {
            int n = intervals.length;

            // sort intervals by their start point
            Arrays.sort(intervals, Comparator.comparingInt(interval -> interval[0]));

            int count = 0;
            int end = intervals[0][1]; // the ending position of the last interval
            for (int i = 1; i < n; i++) {
                if (intervals[i][0] >= end) {
                    // no overlap
                    end = intervals[i][1];
                } else {
                    // overlap
                    count++;
                    // remove a larger interval
                    end = Math.min(end, intervals[i][1]);
                }
            }
            return count;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/NonOverlappingIntervals.kt
package com.github.dkoval.leetcode.challenge

object NonOverlappingIntervalsKt : NonOverlappingIntervals {

    // Resource: https://www.youtube.com/watch?v=BW7LeuJIMhE
    // O(N*logN) time | O(1) space
    override fun eraseOverlapIntervals(intervals: Array<IntArray>): Int {
        if (intervals.size < 2) return 0
        intervals.sortBy { (start, _) -> start }

        var count = 0
        var lastIncludedIntervalIdx = 0
        for (i in 1..intervals.lastIndex) {
            val curr = intervals[i] // current interval
            val last = intervals[lastIncludedIntervalIdx] // last included interval
            if (curr[0] < last[1]) {
                // overlap - either `last` or `curr` interval needs to be removed
                count++
                if (curr[1] < last[1]) {
                    // `last` includes `curr`, therefore remove `last` and set new `last` to `curr`;
                    // otherwise simply remove `curr` (`last` remains the same)
                    lastIncludedIntervalIdx = i
                }
            } else {
                // no overlap - include `curr`
                lastIncludedIntervalIdx = i
            }
        }
        return count
    }
}